nautobot/extras/models/groups.py

```diff
@@ -431,3 +443,348 @@ def clean(self):
 
         # Validate `filter` dict
         self.clean_filter()
+
+    @classmethod
+    def generate_query_for_filter(cls, filter_field, value):
+        """
+        Return a `Q` object generated from the a `filter_field` and `value`.
+
+        :param filter_field:
+            Filter instance
+        :param value:
+            Value passed to the filter
+        """
+        q = models.Q()
+        if isinstance(filter_field, django_filters.MultipleChoiceFilter):
+            for v in value:
+                q |= models.Q(**filter_field.get_filter_predicate(v))
+        else:
+            q |= filter_field.get_filter_predicate(v)
+        return q
+
+    @classmethod
+    def generate_query_for_group(cls, group):
+        """
+        Return a `Q` object generated from all filters for a `group`.
+
+        :param group:
+            DynamicGroup instance
+        """
+        fs = group.filterset_class(group.filter, group.get_queryset())
+        q = models.Q()
+
+        for field_name, value in fs.data.items():
+            filter_field = fs.filters[field_name]
+            q &= cls.generate_query_for_filter(filter_field, value)
+
+        return q
+
+    def generate_members_query(self):
+        """
+        Return a `Q` object generated from all direct members or from self if `filter` is set.
+        """
+        if self.filter:
+            return self.generate_query_for_group(self)
+
+        query = models.Q()
+        for membership in self.dynamic_group_memberships.all():
+            group = membership.group
+            query &= self.generate_query_for_group(group)
```

Comment:
Basically I changed `generate_query()` (previously `process_group_filters()`) to no longer be recursive. We should probably do a deep dive to go over this together. It's not the easiest to follow even in the best case.

Reply:
To answer your question more explicitly, this is ANDing together queries for each group, which is what we want, because they must all match for the queryset to be valid. Each individual group's query is still distinct, but the only way the query works in aggregate is if all conditions match.